src/serializer.cpp

```diff
@@ -144,6 +161,180 @@ prepare() ->
+            auto buf = *mbs.begin();
+            if( buf.size() == 0 )
+            {
+                auto p = mbs.begin();
+                ++p;
+                buf = *p;
+            }
+
+            if( buf.size() < detail::chunk_header_len_ )
+                detail::throw_length_error();
+
+            std::memset(
+                buf.data(), 0x00, detail::chunk_header_len_);
+            tmp0_.commit(detail::chunk_header_len_);
+
+            chunk_header = buf;
+        }
+
+        auto get_output = [&]() -> buffers::mutable_buffer
+        {
+            auto mbs = tmp0_.prepare(tmp0_.capacity());
+            auto buf = *mbs.begin();
+            if( buf.size() == 0 )
+            {
+                auto p = mbs.begin();
+                ++p;
+                buf = *p;
+            }
+
+            if( is_chunked_ )
+            {
+                if( buf.size() <
+                    detail::crlf_len_ + detail::last_chunk_len_ + 1 )
+                    return {};
+
+                auto n =
+                    buf.size() -
+                    detail::crlf_len_ -
+                    detail::last_chunk_len_;
+
+                buf = { buf.data(), n };
+            }
+
+            // BOOST_ASSERT(buf.size() > 0);
+            return buf;
+        };
+
+        auto get_input = [&]() -> buffers::const_buffer
+        {
+            if( st_ == style::buffers )
+            {
+                if( buffers::buffer_size(buf_) == 0 )
+                    return {};
+
+                auto buf = *(buf_.data());
+                BOOST_ASSERT(buf.size() > 0);
+                return buf;
+            }
+            else
+            {
+                if( zbuf.size() == 0 )
+                    return {};
+
+                auto cbs = zbuf.data();
+                auto buf = *cbs.begin();
+                if( buf.size() == 0 )
+                {
+                    auto p = cbs.begin();
+                    ++p;
+                    buf = *p;
+                }
+                BOOST_ASSERT(buf.size() > 0);
+                return buf;
+            }
+        };
+
+        std::size_t num_written = 0;
+        while( true )
+        {
+            auto in = get_input();
+            auto out = get_output();
+            if( out.size() < 6 + 1 )
+            {
+                BOOST_ASSERT(tmp0_.size() > 0);
+                break;
+            }
+
+            auto results = zlib_filter_->on_process(
+                out, in, more_);
+
+            if( st_ == style::buffers )
+            {
+                buf_.consume(results.in_bytes);
+                if( buffers::buffer_size(buf_) == 0 )
+                    more_ = false;
+            }
+            else
+                zbuf.consume(results.in_bytes);
+
+            if( results.out_bytes == 0 )
+                break;
+
+            BOOST_ASSERT(results.out_bytes > 0);
+
+            num_written += results.out_bytes;
+            tmp0_.commit(results.out_bytes);
+        }
```

Comment:
Looking at this loop, I do like that it is kind of generic with the function objects. I think we should explore going all the way and making this loop into a function template whose behavior is customized by template parameters. Or something like that.